kernel/kernel-api/src/test/scala/io/delta/kernel/internal/checksum/ChecksumWriterSuite.scala

```diff
@@ -0,0 +1,198 @@
+  private val PROTOCOL = CRC_FILE_SCHEMA.indexOf("protocol")
+
+  test("write checksum") {
+    val jsonHandler = new MockCheckSumFileJsonWriter()
+    val checksumWriter = new ChecksumWriter(FAKE_DELTA_LOG_PATH)
+    val protocol = createTestProtocol()
+    val metadata = createTestMetadata()
+
+    def testChecksumWrite(txn: Optional[String]): Unit = {
+      val version = 1L
+      val tableSizeBytes = 100L
+      val numFiles = 1L
+
+      checksumWriter.writeCheckSum(
+        mockEngine(jsonHandler = jsonHandler),
+        new CRCInfo(version, metadata, protocol, tableSizeBytes, numFiles, txn)
+      )
+
+      verifyChecksumFile(jsonHandler, version)
+      verifyChecksumContent(jsonHandler.capturedCrcRow, tableSizeBytes, numFiles, txn)
+      verifyMetadataAndProtocol(jsonHandler.capturedCrcRow, metadata, protocol)
+    }
+
+    // Test with and without transaction ID
+    testChecksumWrite(Optional.of("txn"))
+    testChecksumWrite(Optional.empty())
+  }
+
+  private def verifyChecksumFile(jsonHandler: MockCheckSumFileJsonWriter, version: Long): Unit = {
+    assert(jsonHandler.checksumFilePath == s"$FAKE_DELTA_LOG_PATH/${"%020d".format(version)}.crc")
+    assert(jsonHandler.capturedCrcRow.getSchema == CRC_FILE_SCHEMA)
+  }
+
+  private def verifyChecksumContent(
+                                     actualCheckSumRow: Row,
+      expectedTableSizeBytes: Long,
+      expectedNumFiles: Long,
+      expectedTxnId: Optional[String]): Unit = {
+    assert(actualCheckSumRow.getLong(TABLE_SIZE_BYTES) == expectedTableSizeBytes)
+    assert(actualCheckSumRow.getLong(NUM_FILES) == expectedNumFiles)
+    assert(actualCheckSumRow.getLong(NUM_METADATA) == 1L)
+    assert(actualCheckSumRow.getLong(NUM_PROTOCOL) == 1L)
+
+    if (expectedTxnId.isPresent) {
+      assert(actualCheckSumRow.getString(TXN_ID) == expectedTxnId.get())
+    } else {
+      assert(actualCheckSumRow.isNullAt(TXN_ID))
+    }
+  }
+
+  private def verifyMetadataAndProtocol(
+      actualRow: Row,
+      expectedMetadata: Metadata,
+      expectedProtocol: Protocol): Unit = {
+    checkMetadata(expectedMetadata, actualRow.getStruct(METADATA))
+    checkProtocol(expectedProtocol, actualRow.getStruct(PROTOCOL))
+  }
+
+  private def createTestMetadata(): Metadata = {
+    new Metadata(
+      "id",
+      Optional.of("name"),
+      Optional.of("description"),
+      new Format("parquet", Collections.emptyMap()),
+      "schemaString",
+      new StructType(),
+      stringArrayValue(util.Arrays.asList("c3")),
+      Optional.of(123),
+      stringStringMapValue(new util.HashMap[String, String]() {
+        put("delta.appendOnly", "true")
+      })
+    )
+  }
+
+  private def createTestProtocol(): Protocol = {
+    new Protocol(
+      /* minReaderVersion= */ 1,
+      /* minWriterVersion= */ 2,
+      Collections.emptyList(),
+      Collections.emptyList()
+    )
+  }
+
+  private def checkMetadata(expectedMetadata: Metadata, actualMetadataRow: Row): Unit = {
```

Comment:
can we use convert the Row back to Metadata and compare? same for the protocol

Reply:
+1

Or maybe convert them both to `TestRow`s and compare

Reply:
It seems that we need some refactoring here, either implement `compareTo()` on  Metadata or move TestRow from default to internal.

Let create a separate PR for add `compareTo()` to metadata.java but prefer not to block this PR on this